### DISCUSSION TITLE
Minimizing BoolVars and influence on runtime

### DISCUSSION THREAD

**Discussion by DaanSijbring** [google/or-tools#2320] [CP-SAT questions] [answered]
Hi there,

I am modeling a flexible job-shop problem (with the Python CP-SAT solver) that handles an _x_ amount of orders to be planned within their respective deadlines. Now, we have different employees that can perform each of two tasks within an arbitrary order, and each employee has their preference. What I am concerned with right now, is minimizing the workload of **some** of the employees, while others can be freely planned (within their respective limits). 
OptionalIntervalVars are used to determine the intervals for which (overlapping) days a respective employee can perform an individual task of each job. My current approach is to, for each employee who we wish to minimize in workload, select all of the BoolVars indicating the presence of an OptionalIntervalVar, and minimize the sum: 
```
model.Minimize(sum([sum(worker_pres) for worker_pres in totalMinWorkers]))
```

However, depending on the circumstances, this can entail minimizing several hundred presence BoolVars for only twenty or so orders. Also, I put a time limit on the solver because it will run for (at least) over an hour. Is there a smarter way to minimize workload? And one that is more efficient for the solver?

**Comment by lperron** [comment]
Do you use multiple workers in the search ?

**Reply by DaanSijbring** [reply]
Yes, in this case, the employees are the workers. So we have several workers, and for each order and constituting task, the workers able to complete the task are precomputed, along with individual durations.

**Comment by lperron** [comment, marked as answer]
No, I meant use the search parameter: num_workers:8 (or more).
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le mer. 6 janv. 2021 à 17:24, DaanSijbring <notifications@github.com> a
écrit :

**Reply by DaanSijbring** [reply]
That's a useful option... Wow, that really speeds it up! I didn't read about that in the guides I went through, is there any place I could read some more documentation on how this parallelism works? 
Regarding the initial post, do you think it would speed things up if I first summarized (in a BoolVar) whether one of the presences for OptionalIntervalVars for a worker for one task is true? So, for each task and worker I would have a BoolVar indicating if they performed this task, and then minimizing over this list of BoolVars?

**Reply by lperron** [reply]
It depends. You need to try.

**Comment by lperron** [comment]
Yes, it would help.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le mer. 6 janv. 2021 à 18:00, DaanSijbring <notifications@github.com> a
écrit :
